infrastructure/smart-contracts/contracts/upgrades/node-license/NodeLicense8.sol

```diff
@@ -0,0 +1,748 @@
+            _mint(msg.sender, newItemId);
+            _mintTimestamps[newItemId] = block.timestamp;
+        }
+
+        uint16 newAmount = whitelistAmounts[msg.sender] - toMint;
+        whitelistAmounts[msg.sender] = newAmount;
+        emit WhitelistAmountRedeemed(msg.sender, newAmount);
+    }
+
+    /**
+     * @notice Calculates the price for minting NodeLicense tokens.
+     * @param _amount The amount of tokens to mint.
+     * @param _promoCode The promo code to use address.
+     * @return The price in wei.
+     */
+    function price(uint256 _amount, string calldata _promoCode) public view returns (uint256) {
+        uint256 totalSupply = totalSupply();
+        uint256 totalCost = 0;
+        uint256 remaining = _amount;
+        uint256 tierSum = 0;
+
+        for (uint256 i = 0; i < pricingTiers.length; i++) {
+            tierSum += pricingTiers[i].quantity;
+            uint256 availableInThisTier = tierSum > totalSupply
+                ? tierSum - totalSupply
+                : 0;
+
+            if (remaining <= availableInThisTier) {
+                totalCost += remaining * pricingTiers[i].price;
+                remaining = 0;
+                break;
+            } else {
+                totalCost += availableInThisTier * pricingTiers[i].price;
+                remaining -= availableInThisTier;
+                totalSupply += availableInThisTier;
+            }
+        }
+
+        require(remaining == 0, "Not enough licenses available for sale");
+
+        // Apply discount if promo code is active
+        if (_promoCodes[_promoCode].active) {
+            totalCost = totalCost * (100 - referralDiscountPercentage) / 100;
+        }
+
+        return totalCost;
+    }
+
+    /**
+     * @notice Allows a user to claim their referral reward.
+     * @dev The function checks if claiming is enabled and if the caller has a reward to claim.
+     * If both conditions are met, the reward is transferred to the caller and their reward balance is reset.
+     */
+    function claimReferralReward() external {
+        require(claimable, "Claiming of referral rewards is currently disabled");
+        uint256 reward = _referralRewards[msg.sender];
+        require(reward > 0, "No referral reward to claim");
+        _referralRewards[msg.sender] = 0;
+        (bool success, ) = msg.sender.call{value: reward}("");
+        require(success, "Transfer failed.");
+        // Pay Xai & esXAI rewards if they exist
+        uint256 rewardXai = _referralRewardsXai[msg.sender];
+        uint256 rewardEsXai = _referralRewardsEsXai[msg.sender];
+        if(rewardXai > 0){
+            IERC20 token = IERC20(xaiAddress);
+            token.transfer(msg.sender, rewardXai);
+            _referralRewardsXai[msg.sender] = 0;
+        }
+        if(rewardEsXai > 0){
+            IERC20 token = IERC20(esXaiAddress);
+            token.transfer(msg.sender, rewardEsXai);
+            _referralRewardsEsXai[msg.sender] = 0;
+        }
```

Comment:
Currently it seems we are sending the referral amount directly on mint and then increment the amount the user promocode recipient is able to claim afterwards, we should only have 1 of these options right ?

Reply:
I think if you have a second look at these lines, this is inside of the claimReferralReward function which is called by the referral recipient to claim any referral rewards currently stored in the contract.

Reply:
Sorry I misunderstood, this is perfectly fine, thanks for the message :)